text/0517-io-os-reform.md

```diff
@@ -447,7 +458,173 @@ counts, arguments to `main`, and so on).
+  instead eventually introducing a much richer
+  parsing/formatting/(de)serialization framework that works seamlessly
+  with `Reader` and `Writer`.
+
+  Such a framework is out of scope for this RFC, but the
+  endian-sensitive functionality will be provided elsewhere
+  (likely out of tree).
+
+* **The error type**. The traits currently use `IoResult` in their
+  return types, which ties them to `IoError` in particular. Besides
+  being an unnecessary restriction, this type prevents `Reader` and
+  `Writer` (and various adapters built on top of them) from moving to
+  `libcore` -- `IoError` currently requires the `String` type.
+
+  With associated types, there is essentially no downside in making
+  the error type generic.
+
+With those general points out of the way, let's look at the details.
+
+### `Read`
+[Read]: #read
+
+The updated `Reader` trait (and its extension) is as follows:
+
+```rust
+trait Read {
+    type Err; // new associated error type
+
+    // unchanged except for error type
+    fn read(&mut self, buf: &mut [u8]) -> Result<uint, Err>;
+}
+
+// extension trait needed for object safety
+trait ReadExt: Read {
+    fn bytes(&mut self) -> Bytes<Self> { ... }
+    fn chars<'r>(&'r mut self) -> Chars<'r, Self, Err> { ... }
+
+    fn read_to_end(&mut self) -> Result<Vec<u8>, Err> { ... }
+    fn read_to_string(&self) -> Result<String, Err> { ... }
+
+    ... // more to come later in the RFC
+}
+impl<R: Read> ReadExt for R {}
+```
+
+Following the
+[trait naming conventions](https://github.com/rust-lang/rfcs/pull/344),
+the trait is renamed to `Read` reflecting the single method it
+provides.
+
+The `read` method should not involve internal looping (even over
+errors like `EINTR`). It is intended to be atomic.
+
+#### Removed methods
+
+The proposed `Read` trait is much slimmer than today's `Reader`. The vast
+majority of removed methods are parsing/deserialization, which were
+discussed above.
+
+The remaining methods (`read_exact`, `read_at_least`, `push`,
+`push_at_least`) were removed for various reasons:
+
+* `read_exact`, `read_at_least`: these are somewhat more obscure
+  conveniences that are not particularly robust due to lack of
+  atomicity.
+
+* `push`, `push_at_least`: these are special-cases for working with
+  `Vec`, which this RFC proposes to replace with a more general
+  mechanism described next.
+
+To provide some of this functionality in a more composition way,
+extend `Vec<T>` with an unsafe method:
+
+```rust
+unsafe fn with_extra(&mut self, n: uint) -> &mut [T];
+```
+
+This method is equivalent to calling `reserve(n)` and then providing a
+slice to the memory starting just after `len()` entries. Using this
+method, clients of `Read` can easily recover the `push` method.
+
+### `Write`
+[Write]: #write
+
+The `Writer` trait is cut down to even smaller size:
+
+```rust
+trait Write {
+    type Err;
+    fn write(&mut self, buf: &[u8]) -> Result<uint, Err>;
+    fn flush(&mut self) -> Result<(), Err> { ... }
```

Comment:
@kjpgit 

> if you don't want to panic in the drop(), do explicit close.

Note that this is assuming we reverse the policy of not panicking in destructors and we start to do so here. This does feel like it's also diverging from our heavy use of RAII as it basically means that with I/O objects you should *not* rely on RAII and instead call `close()` explicitly. This also propagates outwards to all structs which themselves contain an I/O object (should that wrapper object provide `.close()` and recommend it is not dropped implicitly?

I feel that we must handle the "error on close in `drop`" case gracefully *somehow* as the alternative seems to rail against conventions in Rust elsewhere.

> And you only get one shot at close.

I agree! This does, however, require all I/O objects to have a flag to indicate whether they're closed or not. This "flag" is largely statically determined today (lifetime of the object itself) so it doesn't have to be added everywhere.

> As long as one can wrap a file in a gzwriter, and a AES encryptor, and a
(whatever) and make sure .close() actually works end to end to return an
error

If we were to move `Close` to a trait we could likely add implementations for wrapper objects, but there would unfortunately be no guarantee that it's implemented.

--- 

@untitaker 

> I want such mistakes to be catched at compile-time, not just at runtime.

I alluded to this above a bit, but taken to the extreme I think that this may be very difficult to work with. All types which *contain* an I/O object would be forced to add `.close()` and be warned about as well. We'd also have to deal with the problem of "what if you call `drop` via unwinding"? I think our hand is forced such that we *must* deal with errors on implicit *close* somehow, and the question is largely when and/or if we provide an explicit `close` as well to learn about errors (and whether it's worth it).

---

@nodakai 

> FYI: for the confusion on handling errors from the ill-designed POSIX close(2), see, for example, http://lwn.net/Articles/576478/

Thanks for the link! I had not seen that article before :).